### PR TITLE
[Rust] ray-rs/prototype-v2

### PR DESCRIPTION
**Pull request by jon-chuang**
## Why are these changes needed?

Updated view of Rust Language Worker prototype progress.

How to run the tests
```
cd python
sudo pip install -e . --verbose 

cd ../rust/ray-rs-examples/simple
bazel run test
```

CC: @rkooo567

BCC: @mwtian @ericl 

## Related issue number
replaces: https://github.com/ray-project/ray/pull/21363

Tracking Issue: #20609 

Builds upon:  https://github.com/ray-project/ray/pull/17922, `cpp/src/ray/util`

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/setup.py [hunk 1]
```diff
@@ -149,13 +149,16 @@ def get_packages(self):
 
 if setup_spec.type == SetupType.RAY_CPP:
     setup_spec.files_to_include += ["ray/cpp/default_worker" + exe_suffix]
+    setup_spec.files_to_include += ["ray/rust/rust_worker" + exe_suffix]
```
Comment (jon-chuang) [inline python/setup.py:152]: is this neccessary?

File: rust/ray-rs-sys/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,347 @@
...
+include!(env!("BAZEL_BINDGEN_SOURCE"));
+
+use std::os::raw::*;
+use std::ffi::CString;
+use c_vec::CVec;
+
+#[cfg(not(feature = "bazel"))]
+mod proto;
+
+const ID_ARRAY_LEN: usize = 28;
```
Comment (jon-chuang) [inline rust/ray-rs-sys/src/lib.rs:19]: Get this from proto

File: rust/ray-rs-sys/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,347 @@
...
+            self.as_slice()
+                .iter()
+                .map(|x| format!("{:02x?}", x))
+                .collect::<Vec<_>>()
+                .join(""),
+        )
+    }
+}
+
+impl ObjectID {
```
Comment (jon-chuang) [inline rust/ray-rs-sys/src/lib.rs:38]: could this be replaced by proto methods?

**Comment by bveeramani** [comment]
## :bangbang: ACTION REQUIRED :bangbang:

We've switched our code formatter from YAPF to Black (see #21311).

To prevent issues with merging your code, here's what you'll need to do:

1. Install Black
```
pip install -I black==21.12b0
```

2. Format changed files with Black
```
curl -o format-changed.sh https://gist.githubusercontent.com/bveeramani/42ef0e9e387b755a8a735b084af976f2/raw/7631276790765d555c423b8db2b679fd957b984a/format-changed.sh
chmod +x ./format-changed.sh
./format-changed.sh
rm format-changed.sh
```

3. Commit your changes.
```
git add --all
git commit -m "Format Python code with Black"
```

4. Merge master into your branch.
```
git pull upstream master
```

5. Resolve merge conflicts (if necessary).

After running these steps, you'll have the updated `format.sh`.

**Comment by kfstorm** [comment]
## :bangbang: ACTION REQUIRED :bangbang:

We've updated our formatting configuration for C++ code. (see #22725)

This PR includes C++ code change. To prevent issues with merging your code, here's what you'll need to do:

1. Merge the latest changes from upstream/master branch into your branch.

```bash
git pull upstream master
git merge upstream/master
```

2. Resolve merge conflicts (if necessary).

After running these steps, you'll have the updated C++ formatting configuration.

3. Format changed files.

```bash
scripts/format.sh
```

4. Commit your changes.

```bash
git add --all
git commit -m "Format C++ code"
```

**Comment by crazyboycjr** [comment]
What's the state of this PR? Will ray have rust binding or support in the future?

**Comment by jjyao** [comment]
Hi @crazyboycjr, currently it's not prioritized but in the long term we do want to support different language bindings if there are demands.

**Comment by crazyboycjr** [comment]
Sounds good! Thanks, @jjyao for confirming this!

**Pull request closed without merging**